File: ralib/src/main/java/com/app/ralib/patch/Patch.java
package com.app.ralib.patch;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.nio.file.Files;
import java.nio.file.Path;

public class Patch {
    public static final String TAG = "Patch";

    @NonNull
    public Path patchPath;
    @NonNull
    public PatchManifest manifest;

    @NonNull Path getEntryAssemblyAbsolutePath() {
        return patchPath.resolve(manifest.getEntryAssemblyFile()).toAbsolutePath().normalize();
    }

    @Nullable
    public static Patch fromPatchPath(@NonNull Path patchPath) {
        patchPath = patchPath.normalize();
        if (!Files.exists(patchPath) || !Files.isDirectory(patchPath)) {
            Log.w(TAG, "fromPatchPath: Patch path does not exist or is not a directory: " + patchPath);
            return null;
        }

        var manifest = PatchManifest.fromJson(patchPath.resolve(PatchManifest.MANIFEST_FILE_NAME));
        if (manifest == null) {
            Log.w(TAG, "fromPatchPath: Failed to load manifest from path: " + patchPath);
            return null;
        }

        return new Patch(patchPath, manifest);
    }

    public Patch(@NonNull Path patchPath, @NonNull PatchManifest manifest) {
        this.patchPath = patchPath;
        this.manifest = manifest;
    }
}

File: app/src/main/java/com/app/ralaunch/patch/Patch.kt
package com.app.ralaunch.patch

import android.util.Log
import java.nio.file.Files
import java.nio.file.Path

/**
 * 补丁数据类
 */
data class Patch(
    val patchPath: Path,
    val manifest: PatchManifest
) {
    fun getEntryAssemblyAbsolutePath(): Path {
        return patchPath.resolve(manifest.entryAssemblyFile).toAbsolutePath().normalize()
    }

    companion object {
        private const val TAG = "Patch"

        @JvmStatic
        fun fromPatchPath(patchPath: Path): Patch? {
            val normalizedPath = patchPath.normalize()
            if (!Files.exists(normalizedPath) || !Files.isDirectory(normalizedPath)) {
                Log.w(TAG, "fromPatchPath: Patch path does not exist or is not a directory: $normalizedPath")
                return null
            }

            val manifest = PatchManifest.fromJson(normalizedPath.resolve(PatchManifest.MANIFEST_FILE_NAME))
            if (manifest == null) {
                Log.w(TAG, "fromPatchPath: Failed to load manifest from path: $normalizedPath")
                return null
            }

            return Patch(normalizedPath, manifest)
        }
    }
}
